### PR TITLE
Memory queue: free event data sooner after acknowledgments

### PR DESCRIPTION
**Pull request by faec**
## Proposed commit message

Refactor the memory queue's ackLoop goroutine to allow earlier freeing of event data:
- Previously, each event batch had its own acknowledgment channel, and ackLoop listened to each of them strictly in order. Now, all event batches use a shared acknowledgment channel and ackLoop frees the data for acknowledged events immediately on receiving the signal from that batch, regardless of the order they are received.
- Since acknowledgments no longer happen strictly in queue order, each batch now has a "done" flag owned by the ackLoop goroutine, and this is used to collect acknowledged batches when advancing the queue instead of reading the sequence of acknowledgment channels.

In benchmarks, CPU and ingestion rate were unaffected, while memory used dropped by 0-5% depending on configuration, with the biggest improvements being in the `scale` and `throughput` performance presets.

There is no visible functional difference from this change, except that event pointers are reset to null slightly sooner in the acknowledgment process.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

**Comment by leehinman** [comment]
@faec this might be more trouble than it is worth, but what do you think about adding a benchmark test for this?

I was thinking 2 scenarios might be interesting.  1) single producer and single consumer going as fast as possible.  2) multiple producers going as fast as possible and multiple consumers, but have one of those consumers be "slow".

**Review by belimawr — APPROVED** [review]

**Review by leehinman — APPROVED** [review]
LGTM

**Comment by faec** [comment]
This PR is on hold because https://github.com/elastic/beats/pull/38166 gets much better improvement with a much smaller change, and might make this one completely redundant. Once that one's merged, I'll reevaluate whether this one is still worth going ahead with.

**Comment by faec** [comment]
Closing since this is redundant with other changes that have been merged in the meantime

**Pull request closed without merging**
